http/src/main/java/io/micronaut/http/ssl/SslBuilder.java

```diff
@@ -137,33 +152,269 @@ protected Optional<KeyStore> getKeyStore(SslConfiguration ssl) throws Exception
                             Optional<String> optionalPassword) throws Exception {
-        String type = optionalType.orElse("JKS");
-        String password = optionalPassword.orElse(null);
-        KeyStore store = KeyStore.getInstance(type);
+        return load(new KeyStoreBasedCertificateSpec(optionalType.orElse(null), optionalPassword.orElse(null), null, resource));
+    }
 
-        Optional<URL> url = resourceResolver.getResource(resource);
-        if (url.isPresent()) {
-            store.load(url.get().openStream(), password == null ? null : password.toCharArray());
+    /**
+     * @param spec The configured certificate spec
+     * @return A {@link KeyStore}
+     * @throws Exception if there is an error
+     */
+    protected KeyStore load(CertificateSpec spec) throws Exception {
+        if (spec instanceof KeyStoreBasedCertificateSpec ks) {
+            KeyStore store = createEmptyKeyStore(spec.provider, spec.type == null ? "JKS" : spec.type);
+
+            InputStream stream = resourceResolver.getResourceAsStream(ks.path)
+                .orElseThrow(() -> new SslConfigurationException("The resource " + ks.path + " could not be found"));
+            try {
+                store.load(stream, spec.password == null ? null : spec.password.toCharArray());
+            } catch (IOException e) {
+                if (!(e.getCause() instanceof UnrecoverableKeyException)) {
+                    try {
+                        if (spec.type == null) {
+                            // we can't add passwordless keys to a JKS key store
+                            store = createEmptyKeyStore(spec.provider, "PKCS12");
+                        }
+
+                        loadPem(ks.path, spec.password, spec.provider, store);
+                    } catch (PemParser.NotPemException f) {
+                        // probably should have been loaded as KS
+                        e.addSuppressed(new Exception("Also tried and failed to load the input as PEM:", f));
+                        throw e;
+                    } catch (Exception f) {
+                        // probably should have been loaded as PEM
+                        f.addSuppressed(new Exception("Also tried and failed to load the input as a key store:", e));
+                        throw f;
+                    }
+                } else {
+                    throw e;
+                }
+            }
             return store;
+        } else if (spec instanceof PemBasedCertificateSpec pem) {
+            List<Object> keyItems;
+            try (InputStream s = resourceResolver.getResourceAsStream(pem.keyPath).orElseThrow()) {
+                keyItems = new PemParser(pem.provider, pem.password)
+                    .loadPem(new String(s.readAllBytes(), StandardCharsets.UTF_8));
+            }
+            List<Object> certItems;
+            try (InputStream s = resourceResolver.getResourceAsStream(pem.certificatePath).orElseThrow()) {
```

Comment:
throw a better exception here with a message that includes the certificatePath mentioned it is missing